sig-scheduler-plugins/pkg/controllers/podgroup_controller.go

```diff
@@ -0,0 +1,473 @@
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("reconciling flux-framework/fluence-controller for request")
+	pg := &schedv1alpha1.PodGroup{}
+
+	// Get the timestamp as soon as reconcile happens as a fallback below
+	timestamp := metav1.NewMicroTime(time.Now())
+	if err := r.Get(ctx, req.NamespacedName, pg); err != nil {
+
+		// Case 1: if we get here and it's not found, assume not created
+		if apierrs.IsNotFound(err) {
+			log.Info("PodGroup", "Status", fmt.Sprintf("Pod group %s is not found, deleted.", req.NamespacedName))
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, fmt.Sprintf("Unable to retrieve pod group %s", req.NamespacedName))
+		return ctrl.Result{}, err
+	}
+	log.Info("REFERENCES", "Reconciler", pg.ObjectMeta.OwnerReferences)
+
+	// Grab all statuses (and groups of them) we are interested in
+	schedulingOrPending := (pg.Status.Phase == schedv1alpha1.PodGroupScheduling || pg.Status.Phase == schedv1alpha1.PodGroupPending)
+	twoDaysOld := pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour
+	finishedOrFailed := pg.Status.Phase == schedv1alpha1.PodGroupFinished || pg.Status.Phase == schedv1alpha1.PodGroupFailed
+
+	// Finished or failed - clean up the group
+	if finishedOrFailed {
+		log.Info("PodGroup", "Status", fmt.Sprintf("Pod group %s is finished or failed.", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	// If startScheduleTime - createTime > 2days,
+	// do not reconcile again because pod may have been GCed
+	if schedulingOrPending && pg.Status.Running == 0 && twoDaysOld {
+		r.recorder.Event(pg, v1.EventTypeWarning, "Timeout", "schedule time longer than 48 hours")
+		return ctrl.Result{}, nil
+	}
+
+	// We can get the podList and check for sizes here
+	podList := &v1.PodList{}
+
+	// Select based on the group name
+	groupNameSelector := labels.Set(map[string]string{schedv1alpha1.PodGroupLabel: pg.Name}).AsSelector()
+	err := r.List(ctx, podList, client.MatchingLabelsSelector{Selector: groupNameSelector})
+	if err != nil {
+		log.Error(err, "List pods for group failed")
+		return ctrl.Result{}, err
+	}
+
+	// If the scheduler time created is Zero (not set) we set it here
+	if pg.Status.ScheduleStartTime.IsZero() {
+		return r.setTimeCreated(ctx, pg, podList.Items, timestamp)
+	}
+
+	// Inspect the size, set on the group if not done yet
+	size := len(podList.Items)
+	log.Info("PodGroup", "Name", pg.Name, "Size", size)
+
+	// When first created, size should be unset (MinMember)
+	// Get size label from the first pod
+	if int(pg.Spec.MinMember) == 0 {
+		log.Info("PodGroup", "Status", fmt.Sprintf("Pod group %s updating size to %d", pg.Name, size))
+		return r.updatePodGroupSize(ctx, pg, int32(size), podList.Items)
+
+	} else if int(pg.Spec.MinMember) != size {
```

Comment:
In practice this means that the number of pods we found in the cluster (size) is not equal to the number of pods that the group is expecting. I added this warning specifically for fluence, because we are scheduling groups at _exactly_ a size, and have not yet tested (and thus do not have any kind of guaranteed support) for cases when the size of pods found is greater than `MinMember`, or the `MinMember` argument is treated as it is named - a minimum. I allow it to go through (and only issue a warning) because I think it would still work - but it would assume a more elastic kind of workload. If the number is less than `MinMember` that's not a big deal because fluence won't schedule below that threshold. It will continue / try again. But if it's over that threshold (`MinMember`) what would happen is that fluence would see that MinMember is satisfied, and then allocate nodes. It would only allocate nodes for the size `MinMember`, and those would go into running. The workload could (and would need to be able to) finish at that size, and then be deleted, and the other pods (that were deemed "over what we need") would be deleted, still in pending. This of course would not work for MPI - the subset of pods Running would wait indefinitely for the others they expect to be there (but fluence did not). We have not tested any cases like this yet.

I didn't raise an error because I didn't want someone testing fluence with this case to have it completely fail. I didn't distinguish between less than / greater than because both warrant the same action - issuing a warning to us, the developers, for the time being. To figure out the ideal way to handle this, we need more thinking and testing. There is a lot of undefined behavior in there that I don't understand, and we would need to test this more fully. I put a warning for now because that seems like the best compromise.

Reply:
Thanks for the detailed explanation!